Add Store Update

berezovskyi opened the pull request:
This patch is moved from Gerrit Change 95016.

FIxes #3 

Signed-off-by: Andrew Berezovskyi <andriib@kth.se>

jadelkhoury commented:
Code looks good to me.
The readme.md file is the same as the Store-core?
It would have been nice otherwise with some explanations of the main classes (oslcMessage, Handler, ChangeProvider). Alternatively, a very simple documentation of how it all fits together could be nice.

jadelkhoury commented:
+1 for review

berezovskyi commented:
@jamsden did you have time to look at this?

This patch adds all the remaing "boxes" from the architecture depicted here: https://github.com/eclipse/lyo-store/blob/cb11ca9fd9c01c817b3554e7f197d84f788e8b64/src/store-update/doc/architecture.png: change provider, update manager, and handlers.

berezovskyi commented:
I think we should replace my handmade pub-sub mechanism with http://www.baeldung.com/guava-eventbus if it is able to fulfill all our needs.

berezovskyi commented on src/store-sync/pom.xml:
```diff
@@ -0,0 +1,224 @@
+      <version>${version.lyo}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.eclipse.paho</groupId>
```
should not be here

berezovskyi commented on src/store-sync/README.md:
```diff
@@ -0,0 +1,76 @@
+    <dependency>
+      <groupId>org.eclipse.lyo.tools</groupId>
+      <artifactId>lyo-store</artifactId>
+      <version>0.24.0-SNAPSHOT</version>
```
should be updated

berezovskyi commented on src/store-sync/src/main/java/org/eclipse/lyo/store/sync/StoreUpdateManager.java:
```diff
@@ -0,0 +1,102 @@
+ *
+ * @author Andrew Berezovskyi (andriib@kth.se)
+ * @version $version-stub$
+ * @since 0.0.0
```
since 4.0

berezovskyi commented on src/store-sync/src/main/java/org/eclipse/lyo/store/sync/StoreUpdateManager.java:
```diff
@@ -0,0 +1,102 @@
+package org.eclipse.lyo.store.sync;
```
license

berezovskyi commented:
@jadelkhoury @jamsden shall we add it to the 4.0 release?

berezovskyi commented:
@jadelkhoury @jamsden I hit a snag with the earlier breaking API change where we split TRS log interaction interface in two: `PagedTrs` and `TrsEventHandler`. The latter now requires a Resource to be supplied when it is added/modified/deleted. TRS Sync Handler only has metadata. For now, I have extracted TRS handlers from store-sync into store-sync-trs and disabled the module from the build/deploy.

jadelkhoury commented:
This PR from 2017. I no longer have any ideas what it is about. What's TRS stuff doing under Store?

berezovskyi commented:
> This PR from 2017. I no longer have any ideas what it is about. 

Yes, my memory has faded a bit too. But I think it's about having Store being a cache in front of a non-OSLC app, synchronising the data without making N requests to the API for N requests to the OSLC endpoint. But it was too long ago.

> What's TRS stuff doing under Store?

That's one of the few things I can explain clearly: when a resource is added/modified/deleted through Lyo Store, we add it automatically to the TRS log without having you manually edit it.
